core/dbt/cli/main.py

```diff
@@ -25,6 +28,52 @@ def cli_runner():
     cli()
 
 
+class dbtUsageException(Exception):
+    pass
+
+
+# Programmatic invocation
+class dbtRunner:
+    def __init__(self, project: Project = None, profile: Profile = None):
+        self.project = project
+        self.profile = profile
+
+    def invoke(self, args):
+        dbt_ctx = cli.make_context(cli.name, args)
+        dbt_ctx.obj = {}
+        dbt_ctx.obj["project"] = self.project
+        dbt_ctx.obj["profile"] = self.profile
+
+        try:
+            return cli.invoke(dbt_ctx)
+        except (click.NoSuchOption, click.UsageError) as e:
+            raise dbtUsageException(e.message)
+
+
+# TODO: refactor - consider decorator, or post-init on a dbtContext
+def _initialize_context(ctx):
+    flags = Flags()
+
+    # Tracking
+    initialize_from_flags(flags.ANONYMOUS_USAGE_STATS, flags.PROFILES_DIR)
+    ctx.with_resource(track_run(run_command=flags.WHICH))
+
+    ctx.obj = ctx.obj or {}
+
+    # Profile
+    # TODO: generalize safe access to threads
+    threads = getattr(flags, "THREADS", None)
+    profile = load_profile(flags.PROJECT_DIR, flags.VARS, flags.PROFILE, flags.TARGET, threads)
+    # Project
+    if ctx.obj.get("project") is None:
+        project = load_project(flags.PROJECT_DIR, flags.VERSION_CHECK, profile, flags.VARS)
+        ctx.obj["project"] = project
+
+    # Context for downstream commands
+    ctx.obj["flags"] = flags
```

Comment:
Not saying we need to change it but since we actually construct flags, profile, project within the command, is `ctx.obj` still the best place to put it?

Reply:
Hmmm, definitely worth questioning! An alternative could be to inject profile, project, manifest, etc into the command method's `arg`s, similar to how `click.pass_context` works. This way we'd be a little less coupled to click, although we'd probably still need to use the click context to store any project/profile overrides passed to `dbtRunner`. There would be some naming conflicts to resolve (ex: params from options named `--profile` and a `profile` override).

I think it's worth keeping in mind alternatives to using `ctx.obj` for passing user-configured objects to commands in mind as we go as the current approach doesn't preclude us from changing this up in the future.